### ISSUE TITLE
An InetAddress index incorrectly expects a preceding InetAddressType index

### ISSUE THREAD

**Issue by dataway** [etingof/pysnmp#305]
In pysnmp/smp/mibs/INET-ADDRESS-MIB.py, class InetAddress, methods cloneFromName / cloneAs Name, the code assumes that an index of type InetAddress will be preceded by an InetAddressType index.

This assumption is false. See, for example, [MPLS-VPN-MIB::mplsVpnVrfRouteEntry](http://mibs.snmplabs.com/asn1/MPLS-VPN-MIB):

> mplsVpnVrfRouteEntry OBJECT-TYPE
>    SYNTAX        MplsVpnVrfRouteEntry
>    MAX-ACCESS    not-accessible
>       INDEX  { mplsVpnVrfName, **mplsVpnVrfRouteDest**,
>                mplsVpnVrfRouteMask, mplsVpnVrfRouteTos,
>                mplsVpnVrfRouteNextHop }
>       ::= { mplsVpnVrfRouteTable 1 }
> 
> **mplsVpnVrfRouteDest**  OBJECT-TYPE
>       SYNTAX        **InetAddress**
>       MAX-ACCESS    not-accessible
>       ::= { mplsVpnVrfRouteEntry 1 }

**Comment by etingof** [comment]
My understanding of [rfc4001](https://tools.ietf.org/html/rfc4001#section-4.1) is that `InetAddressType` always goes first in the well-formed MIBs:

```
  When a generic Internet address is used as an index, both the
   InetAddressType and InetAddress objects MUST be used.  The
   InetAddressType object MUST be listed before the InetAddress object
   in the INDEX clause.
```

WDYT?

**Comment by dataway** [comment]
I agree, but the fact is that there are non-compliant MIBs out there which I need to query.

In general I find that pysnmp does not handle non-compliant SNMP as well as net-snmp.

**Comment by etingof** [comment]
Note, that in this case, it is virtually impossible to interpret `mplsVpnVrfRouteDest` correctly because the crucial information on the format of the data is missing...

The way to work-around broken MIBs is this:

* The user fixes broken MIB by hand
* The user compiles fixed MIB into Python with [mibdump](http://snmplabs.com/pysmi/mibdump.html) tool
* The user places compiled MIB into some local directory and points pysnmp to it via [.addMibSource](http://snmplabs.com/pysnmp/docs/api-reference.html#pysnmp.smi.rfc1902.ObjectIdentity.addMibSource)
 
If you wish, I can push fixed Pythonized MIB to [snmplabs](http://mibs.snmplabs.com/pysnmp/) from where it will be pulled automatically if Internet access is available.

**Comment by dataway** [comment]
I don't think it's possible to fix this MIB in this way. The InetAddressType simply doesn't come down the wire as part of the index of `mplsVpnVrfRouteDest`. Instead there's a separate `mplsVpnVrfRouteDestAddrType` entry in the table (which also comes afterwards in an snmpwalk).

I would need to read `mplsVpnVrfRouteDestAddrType` first, then somehow pass the expected address type through when decoding `mplsVpnVrfRouteDest`.

I'm really not sure what the "clean" solution to this problem might be. Is there any way to tell cloneFromName what address type to expect?

**Comment by etingof** [comment]
The absence of `InetAddressType` in the MIB hints that only one (of a few possible) address format would be actually used for `mplsVpnVrfRoute*` components. If that holds, the fix would be to replace `InetAddress` with something more specific in all the affected MIB objects e.g.:

```
mplsVpnVrfRouteDest  OBJECT-TYPE
      SYNTAX        InetAddressIPv6
      MAX-ACCESS    not-accessible
      STATUS        current
      DESCRIPTION
```

Would that work?

**Comment by dataway** [comment]
No, a single table can contain both IPv4 and IPv6 addresses. For each entry there's a separate `mplsVpnVrfRouteDestAddrType` entry that distinguishes the type.

It might be simplest to create a hack-syntax ("InetAddressHack") that creates `InetAddressIPv4` or `InetAddressIPv6` objects depending on whether the length is 4 or 16... but this is clearly a hack.

**Comment by etingof** [comment]
> No, a single table can contain both IPv4 and IPv6 addresses. For each entry there's a separate mplsVpnVrfRouteDestAddrType entry that distinguishes the type.

This looks like chicken-and-egg problem - you can't access the value of `mplsVpnVrfRouteDestAddrType` before index is parsed, is my understanding correct?

> It might be simplest to create a hack-syntax ("InetAddressHack") that creates InetAddressIPv4 or InetAddressIPv6 objects depending on whether the length is 4 or 16... but this is clearly a hack.

That appears like the only way to deal with this problem. Perhaps we could add such hack into existing `InetAddress` as a last resort measure to handle broken indices...?

If you are at it and willing to contribute a patch, that would be awesome! ;-)

**Comment by dataway** [comment]
Ok. I'll send you a patch in the next few days.

**Comment by lextm** [comment]
Fixed with a propery fix and will ship in v7.1.28 in the coming weeks. This can be closed now.
